### PR TITLE
[EmitC] Adapt to lvalue type

### PR DESCRIPTION
**Pull request by simon-camp**
Changes needed to integrate https://github.com/llvm/llvm-project/pull/91475-

### CODE REVIEW

File: compiler/src/iree/compiler/Dialect/VM/Conversion/VMToEmitC/ConvertVMToEmitC.cpp [hunk 1]
```diff
@@ -1249,6 +1275,15 @@ createModuleStructure(IREE::VM::ModuleOp moduleOp,
     emitc_builders::structDefinition(builder, loc, moduleStructStateName,
                                      moduleStructStateFields);
 
+    // Create a typedef for the begin_call member of the `iree_vm_module_t`
+    // type. See `vm/module.h` for the definition. The EmitC dialect doesn't
+    // handle function pointer types currently. Neither can member lookup and
+    // call be fused into one expression at the moment.
```
Comment (marbre) [inline compiler/src/iree/compiler/Dialect/VM/Conversion/VMToEmitC/ConvertVMToEmitC.cpp]: NIT:
```suggestion
    // handle function pointer types currently. Neither member lookups nor
    // calls can be fused into one expression at the moment.
```

File: compiler/src/iree/compiler/Dialect/VM/Conversion/VMToEmitC/ConvertVMToEmitC.cpp [hunk 2]
```diff
@@ -1865,19 +1906,24 @@ class ExportOpConversion : public EmitCConversionPattern<IREE::VM::ExportOp> {
           emitc::PointerType::get(emitc::OpaqueType::get(ctx, argumentsType)),
           /*operand=*/argumentsData);
 
-      argumentStruct.value = arguments.getResult();
+      auto arguments_ =
+          emitc_builders::asLValue(rewriter, loc, arguments.getResult());
+
+      argumentStruct.value = arguments_;
```
Comment (marbre) [inline compiler/src/iree/compiler/Dialect/VM/Conversion/VMToEmitC/ConvertVMToEmitC.cpp]: Maybe change to
```suggestion
      auto argumentsLValue_ =
          emitc_builders::asLValue(rewriter, loc, arguments.getResult());

      argumentStruct.value = argumentsLValue_;
```
for consistency with the other renamings?

File: compiler/src/iree/compiler/Dialect/VM/Conversion/VMToEmitC/ConvertVMToEmitC.cpp [hunk 3]
```diff
@@ -1888,7 +1934,10 @@ class ExportOpConversion : public EmitCConversionPattern<IREE::VM::ExportOp> {
           emitc::PointerType::get(emitc::OpaqueType::get(ctx, resultType)),
           /*operand=*/resultsData);
 
-      resultStruct.value = results.getResult();
+      auto results_ =
+          emitc_builders::asLValue(rewriter, loc, results.getResult());
+
+      resultStruct.value = results_;
```
Comment (marbre) [inline compiler/src/iree/compiler/Dialect/VM/Conversion/VMToEmitC/ConvertVMToEmitC.cpp]: As above:
```suggestion
      auto resultsLValue_ =
          emitc_builders::asLValue(rewriter, loc, results.getResult());

      resultStruct.value = resultsLValue_;
```

**Review by marbre — COMMENTED** [review]
Some first suggestion when going through the conversion. However, I probably need to take a second look.

With the patch applying https://github.com/llvm/llvm-project/commit/e47b507562624bf291ab2515699d39c2669b6131 on top of https://github.com/iree-org/llvm-project/commit/8835909fd7a905712143e3f833b1fb57e5c6d2a7, IREE can be successfully compiled at HEAD (https://github.com/iree-org/iree/commit/5f4235a754a391153db6fe4934dad70159bf26fe) and the relevant tests pass! 🚀

**Review by marbre — APPROVED** [review]
I only checked part of the tests (beside running the tests) but this looks good to me.

File: compiler/src/iree/compiler/Dialect/VM/Conversion/VMToEmitC/ConvertVMToEmitC.cpp [hunk 4]
```diff
@@ -2440,43 +2497,51 @@ class ImportOpConverter {
         Type valueType = llvm::cast<emitc::PointerType>(argType).getPointee();
         Value size =
             emitc_builders::sizeOf(builder, loc, TypeAttr::get(valueType));
-        uint8Ptr = builder
-                       .create<emitc::AddOp>(
-                           /*location=*/loc,
-                           /*type=*/bytePtrType,
-                           /*operands=*/ArrayRef<Value>{uint8Ptr, size})
-                       .getResult();
+        builder
```
Comment (marbre) [inline compiler/src/iree/compiler/Dialect/VM/Conversion/VMToEmitC/ConvertVMToEmitC.cpp]:
```suggestion
        uint8Ptr = builder
```

File: compiler/src/iree/compiler/Dialect/VM/Conversion/VMToEmitC/ConvertVMToEmitC.cpp [hunk 5]
```diff
@@ -2319,21 +2370,26 @@ class ImportOpConverter {
...
-                                                  /*operand=*/arguments);
-
+    auto uint8Ptr = emitc_builders::structMember(builder, loc,
+                                                 /*type=*/bytePtrType,
+                                                 /*memberName=*/"data",
+                                                 /*operand=*/arguments);
     for (size_t i = 0; i < inputTypes.size(); i++) {
       BlockArgument arg = funcOp.getArgument(i + inputOffset);
+      auto argLValue = emitc_builders::asLValue(builder, loc, arg);
+
```
Comment (marbre) [inline compiler/src/iree/compiler/Dialect/VM/Conversion/VMToEmitC/ConvertVMToEmitC.cpp]:
```suggestion
```

File: compiler/src/iree/compiler/Dialect/VM/Conversion/VMToEmitC/ConvertVMToEmitC.cpp [hunk 6]
```diff
@@ -2359,7 +2415,7 @@ class ImportOpConverter {
             emitc_builders::sizeOf(builder, loc, TypeAttr::get(argType));
 
         // memcpy(uint8Ptr, &arg, size);
-        Value argPtr = emitc_builders::addressOf(builder, loc, arg);
+        Value argPtr = emitc_builders::addressOf(builder, loc, argLValue);
```
Comment (marbre) [inline compiler/src/iree/compiler/Dialect/VM/Conversion/VMToEmitC/ConvertVMToEmitC.cpp:2417]:
```suggestion
        auto argLValue = emitc_builders::asLValue(builder, loc, arg);
        Value argPtr = emitc_builders::addressOf(builder, loc, argLValue);
```

File: compiler/src/iree/compiler/Dialect/VM/Conversion/VMToEmitC/ConvertVMToEmitC.cpp (hunk 5 above)
Comment (marbre) [reply compiler/src/iree/compiler/Dialect/VM/Conversion/VMToEmitC/ConvertVMToEmitC.cpp]: Needs to be moved down.

**Comment by simon-camp** [comment]
Landed as part of https://github.com/iree-org/iree/pull/18286

**Pull request closed without merging**